### PR TITLE
Fix massive performance hit due to enabling collision

### LINKED ISSUE #83744: Enabling particle turbulence causes massive performance hit
### Godot version

4.1.2 stable

### System information

macOS, Android. All renderers

### Issue description

After adding a GPUParticles2D node to our project, I was surprised to find it suddenly running at 10FPS on Android. Disabling particle turbulence brought the FPS back to the normal 90. Turns out the real culprit is not the turbulence itself, but this conditional check for collision in the particle shader:

```
if (!COLLIDED) {
		
		float vel_mag = length(VELOCITY);
		float vel_infl = clamp(mix(turbulence_influence_min, turbulence_influence_max, rand_from_seed(alt_seed)) * turbulence_influence, 0.0, 1.0);
		VELOCITY = mix(VELOCITY, normalize(noise_direction) * vel_mag * (1.0 + (1.0 - vel_infl) * 0.2), vel_infl);
	}	
```

This check gets generated even when collision is disabled. Luckily I don't need collision and simply removing it solved the issue (I've included this fixed version of the particles in the MRP). However, it's still odd that this is so slow in the first place. Replacing the if statement with a ternary or a step expression did not help.

The problem is not only on mobile - running the example project on my M1 Mac, I get a drop from 1500 FPS to 600 after enabling turbulence on a default particle system emitting only 1 particle. Also happens regardless of rendering backend. Have not confirmed on iOS or Windows.

I assume it applies to 3D too, but didn't check.

### Steps to reproduce

Run MRP scene, should run well at first.
Then enable turbulence and observe FPS drop.

### Minimal reproduction project

[SlowParticleTurbulenceIssue.zip](https://github.com/godotengine/godot/files/13061515/SlowParticleTurbulenceIssue.zip)

### PR DESCRIPTION
**Pull request by k0T0z**
Fixes #83744 

Do we need to modify this patch #76598?

### CODE REVIEW

**Review by RPicster — APPROVED** [review]

**Comment by RPicster** [comment]
Looks good to me, I also tested the PR and it works as expected.
In my game it even gave a noticable performance increase.

**Comment by k0T0z** [comment]
@RPicster Great, I am gonna add a comment for explanation now.

File: scene/resources/particle_process_material.cpp [hunk 1]
```diff
@@ -984,12 +984,16 @@ void ParticleProcessMaterial::_update_shader() {
...
 		code += "		\n";
 		code += "		vec3 noise_direction = get_noise_direction(TRANSFORM[3].xyz);\n";
-		code += "		if (!COLLIDED) {\n";
-		code += "			\n";
-		code += "			float vel_mag = length(final_velocity);\n";
-		code += "			float vel_infl = clamp(dynamic_params.turb_influence * turbulence_influence, 0.0,1.0);\n";
-		code += "			final_velocity = mix(final_velocity, normalize(noise_direction) * vel_mag * (1.0 + (1.0 - vel_infl) * 0.2), vel_infl);\n";
-		code += "		}\n";
+		// The following snippet causes massive performance hit when collision is disabled, so we don't need it as long as collision is disabled.
+		// Refer to godotengine/godot/issues/83744 for more info.
```
Comment (AThousandShips) [inline scene/resources/particle_process_material.cpp:988]:
```suggestion
		// Refer to https://github.com/godotengine/godot/issues/83744 for more info.
```
Or just `#83744`, half a link isn't very useful IMK

Comment (akien-mga) [reply scene/resources/particle_process_material.cpp:988]: We usually use `GH-83744`.

Comment (k0T0z) [reply scene/resources/particle_process_material.cpp]: @akien-mga @AThousandShips Done

**Comment by k0T0z** [comment]
What I see is that this snippet is a simple calculation going on there, so why there's a massive performance hit happening here?
```c++
if (!COLLIDED) {
	float vel_mag = length(VELOCITY);
	float vel_infl = clamp(mix(turbulence_influence_min, turbulence_influence_max, rand_from_seed(alt_seed)) * turbulence_influence, 0.0, 1.0);
	VELOCITY = mix(VELOCITY, normalize(noise_direction) * vel_mag * (1.0 + (1.0 - vel_infl) * 0.2), vel_infl);
}
```

**Comment by clayjohn** [comment]
> What I see is that this snippet is a simple calculation going on there, so why there's a massive performance hit happening here? Maybe the `rand_from_seed()` function? dk for sure
> 
> ```c++
> if (!COLLIDED) {
> 	float vel_mag = length(VELOCITY);
> 	float vel_infl = clamp(mix(turbulence_influence_min, turbulence_influence_max, rand_from_seed(alt_seed)) * turbulence_influence, 0.0, 1.0);
> 	VELOCITY = mix(VELOCITY, normalize(noise_direction) * vel_mag * (1.0 + (1.0 - vel_infl) * 0.2), vel_infl);
> }
> ```
> 
> EDIT: I noticed the `mix()` function here, I read it `min()`, dk what it means. If someone can explain why this snippet causes these big differences in FPS, I would really appreciate it.

Godot detects when the COLLIDED keyword is used. If it's used anywhere in the shader then Godot will generate the screen space SDF for collisions. The performance cost is not from the shader itself, but from the cost of generating the SDF. If the SDF is used by another effect, then using COLLIDED wouldn't add additional cost

**Comment by k0T0z** [comment]
@clayjohn I noticed ``COLLIDED`` is only used in branches where the collision is enabled, yeah, I get it now thanks for clarifying 👌

**Comment by akien-mga** [comment]
> Do we need to modify this patch #76598?

WDYT @Calinou @RPicster?

**Comment by RPicster** [comment]
I don't know if that was the root of the cause of the original reason for the performance problem. Turbulence will always come at an increased cost. So eventually rather be safe than sorry with the documentation?

**Comment by Calinou** [comment]
> I don't know if that was the root of the cause of the original reason for the performance problem. Turbulence will always come at an increased cost. So eventually rather be safe than sorry with the documentation?

Indeed, we still expect turbulence to take up a significant amount of GPU time on integrated graphics.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**